### PR TITLE
Add cmake flags to control lua include and extern c

### PR DESCRIPTION
**Pull request by SarkisKhachatryan**
* Added cmake variables to control:
      1) `extern "C"` statements, which wraps includes of `lua` headers
      2) which `lua` library to use
      3) include/exclude `googletest` library and tests of `luabind` from overall build phase
* Added some documentation about changes mentioned above

- - -
This change is [Reviewable](https://reviewable.io/reviews/picsart/luabind/1)

### CODE REVIEW

**Review by davits — CHANGES_REQUESTED** [review]
The lua configuration should be done on cmake level. i.e. which lua headers library this library is linked with.

File: include/luabind/lua.hpp [hunk 1]
```diff
@@ -1,10 +1,20 @@
 #ifndef LUABIND_LUA_HPP
 #define LUABIND_LUA_HPP
 
+#if INCLUDE_LUA_LIB
```
Comment (tigrangs) [inline include/luabind/lua.hpp]: lua includes must be done in any case, so this macro is not necessary

**Review by tigrangs — CHANGES_REQUESTED** [review]

File: include/luabind/lua.hpp (hunk 1 above)
Comment (davits) [reply include/luabind/lua.hpp]: second this, lua headers should always be included, however user should be able to configure where lua headers are included from, with cmake configuration.

File: CMakeLists.txt [hunk 2]
```diff
@@ -6,6 +6,16 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
...
+	add_compile_options(-DINCLUDE_LUA_LIB)
+ENDIF(INCLUDE_LUA_LIB)
+set(INCLUDE_LUA_LIB_WITH_EXTERN_C ON CACHE BOOL "flag to determine whether lua library is included with extern 'C'")
+IF(INCLUDE_LUA_LIB_WITH_EXTERN_C)
+	add_compile_options(-DINCLUDE_LUA_LIB_WITH_EXTERN_C)
+ENDIF(INCLUDE_LUA_LIB_WITH_EXTERN_C)
+
 add_subdirectory(third_party)
 
 add_library(luabind INTERFACE)
```
Comment (davits) [inline CMakeLists.txt:22]: `target_link_libraries(luabind INTERFACE lua_headers)` in here, lua_headers should be configurable.

**Review by SarkisKhachatryan — COMMENTED** [review]
changed

> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1#-:-NKIMzoe3RBAX8RiPQCr:b-n9tkhd)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @davits and @tigrangs)

___
*[`CMakeLists.txt` line 21 at r1](https://reviewable.io/reviews/picsart/luabind/1#-NKIMy4mIz1VJzpz6dlr-r1-21:-NKIN2Pn3-BQh-kKS-ff:b-896fix) ([raw file](https://github.com/picsart/luabind/blob/49f8a82e7d3c7f034b14eaa89a5817f842bc5b20/CMakeLists.txt#L21)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

`target_link_libraries(luabind INTERFACE lua_headers)` in here, lua_headers should be configurable.
</blockquote></details>

Done.
___
*[`include/luabind/lua.hpp` line 4 at r1](https://reviewable.io/reviews/picsart/luabind/1#-NKIMy4mIz1VJzpz6dls-r1-4:-NKIN6krDPdBj7iWWNrP:bet0o07) ([raw file](https://github.com/picsart/luabind/blob/49f8a82e7d3c7f034b14eaa89a5817f842bc5b20/include/luabind/lua.hpp#L4)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

second this, lua headers should always be included, however user should be able to configure where lua headers are included from, with cmake configuration.
</blockquote></details>

for now trying to configure headers only specifying names, tested with 2nd lua library, now trying to test with real library

**Review by tigrangs — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1#-:-NKISgwt9im_A36Nmk-Q:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @davits)

**Review by davits — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1)* status: all files reviewed, 3 unresolved discussions (waiting on @SarkisKhachatryan)

___
*[`CMakeLists.txt` line 15 at r2](https://reviewable.io/reviews/picsart/luabind/1#-NKOG2jG8LIiI_d24YfG:-NKOG2jG8LIiI_d24YfH:b71ncnn) ([raw file](https://github.com/picsart/luabind/blob/27fe3839da0512b43f8f46581292264d07c733a8/CMakeLists.txt#L15)):*
> ```CMake
> set(INCLUDE_LUA_LIB_WITH_EXTERN_C ON CACHE BOOL "flag to determine whether lua library is included with extern 'C'")
> IF(INCLUDE_LUA_LIB_WITH_EXTERN_C)
> 	add_compile_options(-DINCLUDE_LUA_LIB_WITH_EXTERN_C)
> ```

Replace <tab> with spaces.
___
*[`include/luabind/lua.hpp` line 4 at r2](https://reviewable.io/reviews/picsart/luabind/1#-NKOGGbdBVLJdULmyDnu:-NKOGGbdBVLJdULmyDnv:b7m7fmu) ([raw file](https://github.com/picsart/luabind/blob/27fe3839da0512b43f8f46581292264d07c733a8/include/luabind/lua.hpp#L4)):*
> ```cpp
> #define LUABIND_LUA_HPP
> 
> #if INCLUDE_LUA_LIB_WITH_EXTERN_C
> ```

#if checks if the macro is defined and has the value of 1.
Use #ifdef instead.

**Review by SarkisKhachatryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1)* status: 1 of 8 files reviewed, 3 unresolved discussions (waiting on @davits and @tigrangs)

___
*[`CMakeLists.txt` line 15 at r2](https://reviewable.io/reviews/picsart/luabind/1#-NKOG2jG8LIiI_d24YfG:-NKP2Xm0A04Ay81v8n8Q:b-896fix) ([raw file](https://github.com/picsart/luabind/blob/27fe3839da0512b43f8f46581292264d07c733a8/CMakeLists.txt#L15)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

Replace <tab> with spaces.
</blockquote></details>

Done.
___
*[`include/luabind/lua.hpp` line 4 at r2](https://reviewable.io/reviews/picsart/luabind/1#-NKOGGbdBVLJdULmyDnu:-NKP2tCbDEFFOdJv38dw:bb7f2mf) ([raw file](https://github.com/picsart/luabind/blob/27fe3839da0512b43f8f46581292264d07c733a8/include/luabind/lua.hpp#L4)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

#if checks if the macro is defined and has the value of 1.
Use #ifdef instead.
</blockquote></details>

done

**Review by davits — CHANGES_REQUESTED** [review]
> Reviewed 1 of 5 files at r2, 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1)* status: all files reviewed, 3 unresolved discussions (waiting on @SarkisKhachatryan)

___
*[`README.md` line 50 at r3](https://reviewable.io/reviews/picsart/luabind/1#-NKsFcGjB-Wjywzda2br:-NKsFcGjB-Wjywzda2bs:b-k1yo2x) ([raw file](https://github.com/picsart/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L50)):*
> ```Markdown
> ## How to install, confugure, build and run
> 
> Ordinary prcedure when using `luabind` as standalone library.
> ```

There is no use for luabind as a standalone library.
This paragraph describes how to do luabind test and development.
___
*[`README.md` line 57 at r3](https://reviewable.io/reviews/picsart/luabind/1#-NKsGHzm2c2Auaewvm3e:-NKsGHzm2c2Auaewvm3f:b-u56ep2) ([raw file](https://github.com/picsart/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L57)):*
> ```Markdown
> git submodule update --recursive --init
> 
> cmake -B build -S .
> ```

I would prefer having this commands as `cmake --preset...`
___
*[`README.md` line 98 at r3](https://reviewable.io/reviews/picsart/luabind/1#-NKsGZtyA1gGSyXdxRV3:-NKsGZtyA1gGSyXdxRV4:bp313u8) ([raw file](https://github.com/picsart/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L98)):*
> ```Markdown
> ```
>     set(INCLUDE_LUA_LIB_WITH_EXTERN_C ON)
>     set(LUA_LIB_NAME lua)
> ```

Actually integrated luabind needs only lua headers. And I would prefer having separate control option for the lua headers rather then `${LUA_LIB_NAME}_headers`. That imposes our naming conventions to the general public, which is not a good thing, however smart they may sound :)

**Review by SarkisKhachatryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1)* status: all files reviewed, 3 unresolved discussions (waiting on @davits)

___
*[`README.md` line 50 at r3](https://reviewable.io/reviews/picsart/luabind/1#-NKsFcGjB-Wjywzda2br:-NLLDFIWEgHU5nZ0Nvmw:bb6un1j) ([raw file](https://github.com/picsart/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L50)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

There is no use for luabind as a standalone library.
This paragraph describes how to do luabind test and development.
</blockquote></details>

Done
___
*[`README.md` line 57 at r3](https://reviewable.io/reviews/picsart/luabind/1#-NKsGHzm2c2Auaewvm3e:-NLt_Y-qCIqJEF56Scbv:buf54y5) ([raw file](https://github.com/picsart/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L57)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

I would prefer having this commands as `cmake --preset...`
</blockquote></details>

Should I add a new preset? Or just change doc.?
___
*[`README.md` line 98 at r3](https://reviewable.io/reviews/picsart/luabind/1#-NKsGZtyA1gGSyXdxRV3:-NLzSPJbBFsRCmvR7wNF:b-h4yb37) ([raw file](https://github.com/picsart/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L98)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

Actually integrated luabind needs only lua headers. And I would prefer having separate control option for the lua headers rather then `${LUA_LIB_NAME}_headers`. That imposes our naming conventions to the general public, which is not a good thing, however smart they may sound :)
</blockquote></details>

```
set(LUA_LIB_NAME luabind_lua CACHE STRING "lua library source names")
set(LUA_LIB_HEADERS luabind_lua_headers CACHE STRING "lua library header names")
```

I guess you request something like this?

Please note that `LUA_LIB_NAME` is used in test section

**Review by davits — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1)* status: all files reviewed, 3 unresolved discussions (waiting on @SarkisKhachatryan)

___
*[`README.md` line 57 at r3](https://reviewable.io/reviews/picsart/luabind/1#-NKsGHzm2c2Auaewvm3e:-NLzqOQj6v0LvKSCoUNe:b42uj4t) ([raw file](https://github.com/picsart/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L57)):*
<details><summary><i>Previously, SarkisKhachatryan (Sargis Khachatryan) wrote…</i></summary><blockquote>

Should I add a new preset? Or just change doc.?
</blockquote></details>

Presets are there, just test them and change the doc.
___
*[`README.md` line 98 at r3](https://reviewable.io/reviews/picsart/luabind/1#-NKsGZtyA1gGSyXdxRV3:-NLzuNvL-Y2veENIHUr1:b-bq118u) ([raw file](https://github.com/picsart/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L98)):*
<details><summary><i>Previously, SarkisKhachatryan (Sargis Khachatryan) wrote…</i></summary><blockquote>

```
set(LUA_LIB_NAME luabind_lua CACHE STRING "lua library source names")
set(LUA_LIB_HEADERS luabind_lua_headers CACHE STRING "lua library header names")
```

I guess you request something like this?

Please note that `LUA_LIB_NAME` is used in test section
</blockquote></details>

```
set(LUA_LIB_NAME luabind_lua CACHE STRING "lua library source names")
set(LUA_LIB_HEADERS ${LUA_LIB_NAME}_headers CACHE STRING "lua library header names")
```
More like this.
But on the second thought. Now I am doubting that we will ever need link luabind with headers, we can link directly with lua instead and get rid of all this.

**Review by SarkisKhachatryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/PicsArt/luabind/1)* status: 6 of 10 files reviewed, 3 unresolved discussions (waiting on @davits and @tigrangs)

___
*[`README.md` line 57 at r3](https://reviewable.io/reviews/PicsArt/luabind/1#-NKsGHzm2c2Auaewvm3e:-NM3EIxV9iK2fX_pZAon:b-896fix) ([raw file](https://github.com/PicsArt/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L57)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

Presets are there, just test them and change the doc.
</blockquote></details>

Done.
___
*[`README.md` line 98 at r3](https://reviewable.io/reviews/PicsArt/luabind/1#-NKsGZtyA1gGSyXdxRV3:-NM3GIsW1o7mzXbY4xKl:b-896fix) ([raw file](https://github.com/PicsArt/luabind/blob/e64b35662a3e115ea96d1660d611f57d7e2e3174/README.md#L98)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

```
set(LUA_LIB_NAME luabind_lua CACHE STRING "lua library source names")
set(LUA_LIB_HEADERS ${LUA_LIB_NAME}_headers CACHE STRING "lua library header names")
```
More like this.
But on the second thought. Now I am doubting that we will ever need link luabind with headers, we can link directly with lua instead and get rid of all this.
</blockquote></details>

Done.

**Review by davits — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1)* status: all files reviewed, 1 unresolved discussion (waiting on @SarkisKhachatryan)

___
*[`.gitignore` line 3 at r4](https://reviewable.io/reviews/picsart/luabind/1#-NM3OWfMCY6h8Jds1Kun:-NM3OWfMCY6h8Jds1Kuo:bki71fo) ([raw file](https://github.com/picsart/luabind/blob/a32823f082f7b898b709257bd259a7314f97de19/.gitignore#L3)):*
> ```ignore
> .cache
> build/
> CMakeUserPresets.json
> ```

Why? It is a part of build process and VSCode integration...

**Review by SarkisKhachatryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/PicsArt/luabind/1)* status: all files reviewed, 1 unresolved discussion (waiting on @davits)

___
*[`.gitignore` line 3 at r4](https://reviewable.io/reviews/PicsArt/luabind/1#-NM3OWfMCY6h8Jds1Kun:-NM3PS4cC_gbpelPvzjZ:bda1aih) ([raw file](https://github.com/PicsArt/luabind/blob/a32823f082f7b898b709257bd259a7314f97de19/.gitignore#L3)):*
<details><summary><i>Previously, davits (Davit Samvelyan) wrote…</i></summary><blockquote>

Why? It is a part of build process and VSCode integration...
</blockquote></details>

From CMake documentation.
```
CMakePresets.json may be checked into a version control system, and CMakeUserPresets.json should NOT be checked in. For example, if a project is using Git, CMakePresets.json may be tracked, and CMakeUserPresets.json should be added to the .gitignore.
```
just in case )

**Comment by davits** [comment]
*[`.gitignore` line 3 at r4](https://reviewable.io/reviews/PicsArt/luabind/1#-NM3OWfMCY6h8Jds1Kun:-NM3Pzo5BoDIrNyeJ4Kg:bbik28d) ([raw file](https://github.com/PicsArt/luabind/blob/a32823f082f7b898b709257bd259a7314f97de19/.gitignore#L3)):*
<details><summary><i>Previously, SarkisKhachatryan (Sargis Khachatryan) wrote…</i></summary><blockquote>

From CMake documentation.
```
CMakePresets.json may be checked into a version control system, and CMakeUserPresets.json should NOT be checked in. For example, if a project is using Git, CMakePresets.json may be tracked, and CMakeUserPresets.json should be added to the .gitignore.
```
just in case )
</blockquote></details>

Sorry false alarm. Didn't notice User in the name :)

**Review by davits — COMMENTED** [review]
Welcome to the Open Source world! :tada: 

> *[Reviewable](https://reviewable.io/reviews/PicsArt/luabind/1#-:-NM3Q3VBBtiP5nQxbfk-:bc863rt)* status: all files reviewed, 1 unresolved discussion (waiting on @SarkisKhachatryan)

**Review by davits — APPROVED** [review]
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/picsart/luabind/1)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @SarkisKhachatryan)

**Pull request merged**